contracts/adapters/Distribute.sol

```diff
@@ -0,0 +1,277 @@
+                IVoting.VotingState.PASS,
+            "proposal did not pass"
+        );
+
+        distribution.status = GuildKickStatus.IN_PROGRESS;
+        distribution.blockNumber = block.number;
+        ongoingDistributions[address(dao)] = proposalId;
+    }
+
+    /**
+     * @notice Transfers the funds from the Guild account to the member's internal accounts.
+     * @notice The amount of funds is caculated using the historical number of shares of each member.
+     * @dev A distribution proposal must be in progress.
+     * @dev Only proposals that have passed the voting can be completed.
+     * @dev Only active members can receive funds.
+     * @param dao The dao address.
+     * @param toIndex The index to control the cached for-loop.
+     */
+    function distribute(DaoRegistry dao, uint256 toIndex) external override {
+        // Checks if the proposal does not exist or is not completed yet
+        bytes32 ongoingProposalId = ongoingDistributions[address(dao)];
+        Distribution storage distribution = kicks[address(dao)][ongoingProposalId];
+        uint256 blockNumber = distribution.blockNumber;
+        require(
+            distribution.status == DistributionStatus.IN_PROGRESS,
+            "distribution completed or does not exist"
+        );
+
+        // Check if the given index was already processed
+        uint256 currentIndex = distribution.currentIndex;
+        require(currentIndex <= toIndex, "toIndex too low");
+
+        // Get the total number of shares when the proposal was processed.
+        BankExtension bank = BankExtension(dao.getExtensionAddress(BANK));
+        uint256 totalShares = bank.getPriorAmount(TOTAL, SHARES, blockNumber);
+
+        if (distribution.shareHolderAddr != address(0x0)) {
+            // Distributes the funds to 1 share holder only
+            _distribute(dao, distribution.shareHolderAddr, token, amount, totalShares, blockNumber);
+            distribution.status = DistributionStatus.DONE;
+            //emit event
+        } else {
+            // Set the max index supported which is based on the number of members
+            uint256 nbMembers = dao.getNbMembers();
+            uint256 maxIndex = toIndex;
+            if (maxIndex > nbMembers) {
+                maxIndex = nbMembers;
+            }
+            // Distributes the funds to all share holders / active members of the DAO.
+            for (uint256 i = currentIndex; i < maxIndex; i++) {
+                _distribute(dao, dao.getMemberAddress(i), token, amount, totalShares, blockNumber);
+            }
+            distribution.currentIndex = maxIndex;
+            if (maxIndex == nbMembers) {
+                distribution.status = DistributionStatus.DONE;
+                //emit event
+            }
+        }
+        
+    }
+
+    /** 
+     * @notice Transfers the funds from the internal Guild account to the internal member's account.
+     */
+    function _distribute(
```

Comment:
I see that the fair share is calculated here, in case of distribution of EVERY member that is correct,
but if we distribute to a particular member, the member should get the amount (not the fair share of that amount)